### PR TITLE
Move The 'wait' Argument Handling Logic To Its Exec Function

### LINKED ISSUE #1470: Template does not evaluate as expected in the pods field of KubeExecAll
**Describe the bug**
pods field in KubeExecAll can be go-templated. Despite my template is correct given the postgresql object I pass to the action set Kanister can't interpret it.

```
pods: "{{range until (int .Object.spec.numberOfInstances)}} {{ $.Object.metadata.name }}-{{.}}{{end}}"
```

generate this error message 
```
 error:
    message: |-
      pods ".Object.spec.numberOfInstances)}}" not found
      pods "until" not found
      pods "}}-{{.}}{{end}}" not found
      pods "{{range" not found
      pods "$.Object.metadata.name" not found
      pods "(int" not found
      pods "{{" not found
  state: failed
```

For instance this template 
```
pods: '{{range until 3}} acid-pg-cluster-{{.}}{{end}}'
```
or  this one 
```
pods: "{{ .Object.metadata.name }}-0 {{ .Object.metadata.name }}-1 {{ .Object.metadata.name }}-2"
```
 work fine.

I also check the validity of this template with this go code 
```
package main

import (
	"fmt"
	"os"
	"text/template"

	"github.com/Masterminds/sprig"
)

func main() {

	// Set up variables and template.
	numberOfInstances := int64(3)
	vars := map[string]interface{}{
		"Object": map[string]interface{}{
			"metadata": map[string]interface{}{"name": "acid-pg-cluster"},
			"spec":     map[string]interface{}{"numberOfInstances": numberOfInstances}}}
	tpl := `{{range until (int .Object.spec.numberOfInstances)}} {{ $.Object.metadata.name }}-{{.}}{{end}}`

	// Get the Sprig function map.
	fmap := sprig.TxtFuncMap()
	t := template.Must(template.New("test").Funcs(fmap).Parse(tpl))

	err := t.Execute(os.Stdout, vars)
	if err != nil {
		fmt.Printf("Error during template execution: %s", err)
		return
	}

}
```

**To Reproduce**

Install zalando operator and a postgresql example 
```
helm repo add postgres-operator-charts https://opensource.zalando.com/postgres-operator/charts/postgres-operator
helm install postgres-operator postgres-operator-charts/postgres-operator
cat <<EOF | kubectl create -f -
apiVersion: "acid.zalan.do/v1"
kind: postgresql
metadata:
  name: acid-pg-cluster
  namespace: "default"
  labels:
    team: acid
spec:
  teamId: "acid"
  volume:
    size: "200Gi"
  numberOfInstances: 3
  enableMasterLoadBalancer: false
  enableReplicaLoadBalancer: false
  allowedSourceRanges:  # load balancers' source ranges for both master and replica services
  - 127.0.0.1/32
  postgresql:
    version: "13"
    parameters: # Expert section
      shared_buffers: "500MB"
      effective_cache_size: "1500MB"
      max_connections: "300" #setting this based on mariadb (cmdb)
      log_statement: "ddl"
      wal_compression: "on"
      work_mem: "40MB"
      random_page_cost: "1.0"
      autovacuum_vacuum_scale_factor: "0.1"
      logging_collector: "off"
      log_destination: "stderr"
      shared_preload_libraries: "pgaudit,pg_stat_statements,pg_stat_kcache"
      wal_keep_size: "1024MB"
  enableShmVolume: false
  # podPriorityClassName: "nom-backend-service-db"
  spiloRunAsUser: 101
  spiloFSGroup: 103
  resources:
    requests:
      cpu: "1000m"
      memory: "1500Mi"
    limits:
      cpu: "2000m"
      memory: "1500Mi"
  patroni:
    initdb:
      encoding: "UTF8"
      locale: "en_US.UTF-8"
      data-checksums: "true"
    #pg_hba, please note: whole pg_hba.conf gets overwritten.
    ttl: 30
    loop_wait: &loop_wait 10
    retry_timeout: 10
    synchronous_mode: true
    synchronous_mode_strict: true
    maximum_lag_on_failover: 33554432    
  serviceAnnotations:
    prometheus.io/path: /metrics
    prometheus.io/port: "9187"
    prometheus.io/scrape: "true"
EOF
```

Wait for the acid-pg-cluster to be ready then install this blueprint 
```
cat <<EOF | kubectl create -f -
apiVersion: cr.kanister.io/v1alpha1
kind: Blueprint
metadata: 
  name: zalando-bp-simplified
  namespace: kasten-io  
actions:
  helloZalando:
    phases:    
    - func: KubeExecAll
      name: helloToAllPods      
      args:
        namespace: "{{ .Object.metadata.namespace }}"
        # work
        #pods: '{{range until 3}} acid-pg-cluster-{{.}}{{end}}'
        # does not work we're investigating
        pods: "{{range until (int .Object.spec.numberOfInstances)}} {{ $.Object.metadata.name }}-{{.}}{{end}}"
        # work
        #pods: "{{ .Object.metadata.name }}-0 {{ .Object.metadata.name }}-1 {{ .Object.metadata.name }}-2"
        containers: postgres
        command:
        - bash
        - -o
        - errexit
        - -o
        - pipefail
        - -c
        - |
           touch helloZalando.txt            
EOF
```

And execute the actionset  
```
cat <<EOF | kubectl create -f -
apiVersion: cr.kanister.io/v1alpha1
kind: ActionSet
metadata:  
  name: hello-zalando-actionset
  namespace: kasten-io  
spec:
  actions:
  - blueprint: zalando-bp-simplified
    name: helloZalando
    object:
      apiVersion: v1
      group: acid.zalan.do
      kind: ""
      name: acid-pg-cluster
      namespace: default
      resource: postgresqls
    preferredVersion: v1.0.0-alpha
    profile:
      apiVersion: v1alpha1
      group: ""
      kind: profile
      name: kanister-phase-k10-location-k7f9x
      namespace: kasten-io
      resource: ""
EOF
```

**Expected behavior**
Actionset should complete but get the error described above

**Environment**
Kubernetes Version/Provider: 
```
k version 
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.1-5-g76a04fc", GitCommit:"c66c03f3012a10f16eb86fdce6330433adf6c9ee", GitTreeState:"clean", BuildDate:"2021-03-13T02:23:52Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.6", GitCommit:"07959215dd83b4ae6317b33c824f845abd578642", GitTreeState:"clean", BuildDate:"2022-03-30T18:28:25Z", GoVersion:"go1.16.12", Compiler:"gc", Platform:"linux/amd64"}
```
Storage Provider: azure
Cluster Size (#nodes): 3
Data Size: 200Gi 
kasten version : 4.5.14

### PR DESCRIPTION
**Pull request by ihcsim**
## Change Overview

This PR moves the handling of the `wait` function's arguments to its `Exec()` function, to give it more control over the rendering of the `conditions` arguments, which are made up of a mixed of Go templates with nested jsonpath variables. This change isolates the rendering of these `conditions` to the `wait` function code, allowing the removal of the jsonpath check from the general `param.render()` function.

The previous implementation of rendering Go templates with nested jsonpath variables doesn't work with templates contains control structure like conditionals and loops. These constructs may refer to root-scoped variables using the `$` reference, leading to template validation errors, when these root-scoped variables are being misinterpreted as jsonpath variables by the [existing regex](https://github.com/kanisterio/kanister/blob/c1069383ed727248c47126f3b4a7dbeebc62e88c/pkg/jsonpath/jsonpath.go#L27). Since Go regex [doesn't support negative lookbehind](https://groups.google.com/g/golang-nuts/c/Z82JzvhPiHA), there isn't an easy way to update the regex to match "only one `{` and `}` but not before or after". 

The handling of all other arguments remain unchanged.

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [ ] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [x] :bug: Bugfix
- [ ] :sunflower: Feature
- [ ] :world_map: Documentation
- [ ] :robot: Test

## Issues 

- fixes #1470

## Test Plan

Deploy this Nginx statefulset workload to your cluster:

```sh
cat <<EOF apiVersion: apps/v1 | kubectl apply -f -
kind: StatefulSet
metadata:
  name: nginx
  labels:
    app: nginx
spec:
  serviceName: nginx
  replicas: 3
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx
  volumeClaimTemplates:
  - metadata:
      name: www
    spec:
      accessModes: ["ReadWriteOnce"]
      storageClassName: local-path      # update this to match your setup
      resources:
        requests:
          storage: 1Gi
EOF
```

Deploy this blueprint to test the `KubeExecAll` and `Wait` functions:

```sh
cat<<EOF | kubectl apply -f -
apiVersion: cr.kanister.io/v1alpha1
kind: Blueprint
metadata:
  name: exec-all
  namespace: kanister
actions:
  execAll:
    phases:
    - func: KubeExecAll
      name: execAll
      args:
        namespace: "{{ .Object.metadata.namespace }}"
        pods: "{{range until (int .Object.spec.replicas)}} {{ $.Object.metadata.name }}-{{.}}{{end}}"   # should work with range loop with root-scoped variable
        containers: nginx
        command:
        - sh
        - -c
        - |
          echo "hello world"
  execAllErr:
    phases:
    - func: KubeExecAll
      name: execAllErr
      args:
        namespace: "{{ .Object.metadata.namespace }}"
        pods: "{{range until (int .Object.spec.replicas)}} {{ $.Foo.metadata.name }}-{{.}}{{end}}"   # should catch invalid Foo field
        containers: nginx
        command:
        - sh
        - -c
        - |
          echo "hello world"
  checkReady:
    phases:
    - func: Wait
      name: checkReady
      args:
        timeout: 60s
        conditions:
          allOf:
          - condition: '{{ if (eq "{ $.status.readyReplicas }" "3")}}true{{ else }}false{{ end }}'  # should still work with jsonpath variables in wait conditions
            objectReference:
              apiVersion: v1
              group: apps
              resource: statefulsets
              name: "{{ .Object.metadata.name }}"
              namespace: "{{ .Object.metadata.namespace }}"
EOF
```

Deploy this actionset to trigger the `execAll` action and expect the actionset to complete successfully:

```sh
cat <<EOF | kubectl create -f -
apiVersion: cr.kanister.io/v1alpha1
kind: ActionSet
metadata:
  generateName: exec-all-
  namespace: kanister
spec:
  actions:
  - name: execAll
    blueprint: exec-all
    object:
      kind: StatefulSet
      name: nginx
      namespace: default
EOF
```

Deploy this actionset to trigger the `execAllErr` action and expect the actionset to fail and report the unidenfitied `Foo` field:

```sh
cat <<EOF | kubectl create -f -
apiVersion: cr.kanister.io/v1alpha1
kind: ActionSet
metadata:
  generateName: exec-all-err-
  namespace: kanister
spec:
  actions:
  - name: execAll
    blueprint: exec-all
    object:
      kind: StatefulSet
      name: nginx
      namespace: default
EOF
```

Deploy this actionset to trigger the `wait` action and expect the actionset to complete successfully:

```sh
cat<<EOF | kubectl apply -f -
apiVersion: cr.kanister.io/v1alpha1
kind: ActionSet
metadata:
  generateName: wait-
  namespace: kanister
spec:
  actions:
  - name: checkReady
    blueprint: exec-all
    object:
      kind: StatefulSet
      name: nginx
      namespace: default
EOF
```

- [x] :muscle: Manual
- [x] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

File: pkg/phase.go [hunk 1]
```diff
@@ -84,6 +87,18 @@ func (p *Phase) Exec(ctx context.Context, bp crv1alpha1.Blueprint, action string
 	return p.f.Exec(ctx, tp, p.args)
 }
 
+func renderFuncArgs(
+	funcName string,
+	args map[string]interface{},
+	tp param.TemplateParams) (map[string]interface{}, error) {
+	// let wait handle its own go template + jsonpath mixed arguments
+	if strings.ToLower(funcName) == "wait" {
```
Comment (PrasadG193) [inline pkg/phase.go:95]: Can we use the declared name from function pkg - `function.WaitFuncName`

File: pkg/function/wait.go [hunk 2]
```diff
@@ -69,15 +69,23 @@ func (*waitFunc) Name() string {
...
+	rendered, err := param.RenderArgs(args, tp)
+	if err != nil {
+		return nil, err
+	}
+
 	var timeout string
-	var conditions WaitConditions
-	var err error
-	if err = Arg(args, WaitTimeoutArg, &timeout); err != nil {
+	if err := Arg(rendered, WaitTimeoutArg, &timeout); err != nil {
```
Comment (PrasadG193) [inline pkg/function/wait.go:78]: Why do we need rendered arg to parse timeout? As per my understanding, it's always going to be const value.

File: pkg/function/wait.go [hunk 3]
```diff
@@ -144,8 +152,17 @@ func waitForCondition(ctx context.Context, dynCli dynamic.Interface, waitCond Wa
...
+func evaluateCondition(ctx context.Context, dynCli dynamic.Interface, cond Condition, tp param.TemplateParams) (bool, error) {
+	objRefRaw := map[string]crv1alpha1.ObjectReference{
+		"objRef": cond.ObjectReference,
+	}
+	rendered, err := param.RenderObjectRefs(objRefRaw, tp)
+	if err != nil {
+		return false, err
+	}
+	objRef := rendered["objRef"]
+
```
Comment (PrasadG193) [inline pkg/function/wait.go:164]: Should we move this to a separate helper function?

**Comment by michaelcourcy** [comment]
@ihcsim I've been testing on the couchbase cluster example this blueprint 

```
apiVersion: cr.kanister.io/v1alpha1
kind: Blueprint
metadata:
  # test PR 1630 https://github.com/kanisterio/kanister/pull/1630
  name: couchbase-blueprint-test-1630
  namespace: kasten-io
actions:
  backup:
    kind: CustomResource
    phases:
    - func: KubeExecAll
      name: testExecAll      
      args:
        namespace: "{{ .Object.metadata.namespace }}"
        containers: couchbase-server
        pods: "{{range until (int (index .Object.spec.servers 0).size )}} {{ $.Object.metadata.name }}-000{{.}}{{end}}"
        command:
          - bash
          - -o
          - pipefail
          - -o
          - errexit
          - -c
          - |
            touch /tmp/execAll-1630
    - func: Wait
      name: waitCouchbaseClusterReadyAndBalanced
      args:
        timeout: 60s
        conditions:
          allOf:
            - condition: '{{$available := false}}{{ range $condition := $.status.conditions }}{{ if and (eq .type "Available") (eq .status "True")  }}{{$available=true}}{{ end }}{{ end }}{{ $available }}'
              objectReference:
                apiVersion: couchbase.com/v2
                resource: couchbaseclusters
                name: "{{ .Object.Name }}"
                namespace: "{{ .Object.namespace}}"
            - condition: '{{$balanced := false}}{{ range $condition := $.status.conditions }}{{ if and (eq .type "Balanced") (eq .status "True")  }}{{$balanced=true}}{{ end }}{{ end }}{{ $balanced }}'
              objectReference:
                apiVersion: couchbase.com/v2
                resource: couchbaseclusters
                name: "{{ .Object.Name }}"
                namespace: "{{ .Object.namespace}}"
```

The first func kubeExecAll work as expected but the second Wait lead to this error 
```
"template: config:1:47: executing \"config\" at <$.status.conditions>:
                  can't evaluate field status in type param.TemplateParams"
```

In your example I can see that you use `{ ` and ` }` inside a go template expression 

```
condition: '{{ if (eq "{ $.status.readyReplicas }" "3")}}true{{ else }}false{{ end }}'  # should still work with jsonpath variables in wait conditions
```

I'm not sure I understand why you are using that, is it to evaluate a jsonpath expression or the contrary avoid the evaluation as a jsonpath ?

File: pkg/phase.go (hunk 1 above)
Comment (ihcsim) [reply pkg/phase.go:95]: No, it will lead to an import loop.

**Comment by ihcsim** [comment]
@michaelcourcy thanks for trying it out. I thought the proper jsonpath syntax requires the `{` and `}` delimiter, i.e. `${.status.conditions}`. See original code [here](https://github.com/kanisterio/kanister/blob/c1069383ed727248c47126f3b4a7dbeebc62e88c/pkg/jsonpath/jsonpath.go#L27).

File: pkg/function/wait.go (hunk 2 above)
Comment (ihcsim) [reply pkg/function/wait.go:78]: I prefer to retain the original behaviour where we used the rendered argument in `Arg()`, in this PR. I am curious - do we support Go template everywhere in the blueprint, or only specific fields?

**Comment by PrasadG193** [comment]
I am working on switching to go-template syntax for fetching k8s object value in wait condition. Can we hold off on the PR till then?

**Comment by michaelcourcy** [comment]
> I am working on switching to go-template syntax for fetching k8s object value in wait condition. Can we hold off on the PR till then?

I really would prefer that personally, it's a bit ubiquitous to switch from jsonpath to go-template in the same expression.

**Comment by PrasadG193** [comment]
I have raised - https://github.com/kanisterio/kanister/pull/1635 on top of the current PR

**Comment by PrasadG193** [comment]
Since we decided to implement a new function which supports only `go-template` syntax. This PR needs to be refactored to remove the changes in wait function implementation to fix only #1470

**Comment by pavannd1** [comment]
@PrasadG193 I have approved this. Do you want to merge this and refactor the required things in your PR?

**Pull request merged**